pandas/core/common.py

```diff
@@ -1374,36 +1374,44 @@ def _fill_zeros(result, x, y, name, fill):
     mask the nan's from x
     """
 
-    if fill is not None:
+    if fill is None or is_float_dtype(result):
+        return result
+
+    if name.startswith(('r', '__r')):
+        x,y = y,x
 
-        if name.startswith('r'):
-            x,y = y,x
+    if not isinstance(y, np.ndarray):
+        dtype, value = _infer_dtype_from_scalar(y)
+        y = np.empty(result.shape, dtype=dtype)
+        y.fill(value)
 
+    if is_integer_dtype(y):
 
-        if not isinstance(y, np.ndarray):
-            dtype, value = _infer_dtype_from_scalar(y)
-            y = np.empty(result.shape, dtype=dtype)
-            y.fill(value)
+        if (y == 0).any():
 
-        if is_integer_dtype(y):
+            # GH 7325, mask and nans must be broadcastable (also: PR 9308)
+            # Raveling and then reshaping makes np.putmask faster
+            mask = ((y == 0) & ~np.isnan(result)).ravel()
 
-            if (y.ravel() == 0).any():
-                shape = result.shape
-                result = result.ravel().astype('float64')
+            shape = result.shape
+            result = result.ravel().astype('float64', copy=False)
 
-                # GH 7325, mask and nans must be broadcastable
-                signs = np.sign(result)
-                mask = ((y == 0) & ~np.isnan(x)).ravel()
+            np.putmask(result, mask, fill)
 
-                np.putmask(result, mask, fill)
+            # if we have a fill of inf, then sign it correctly
+            # (GH 6178 and PR 9308)
+            if np.isinf(fill):
+                if name.startswith(('r', '__r')):
+                    signs = np.sign(y).ravel()
```

Comment:
Are you sure this improves readability?  (of course, performance will be the same)

The ternary operator is [frowned upon by some Pythonistas](http://stackoverflow.com/a/394814/2223706).